File: repertoire.server/api/requests/song_section_requests.go
package requests

import "github.com/google/uuid"

type GetSongSectionsRequest struct {
	SongID uuid.UUID `form:"songId" validate:"required"`
}

type CreateSongSectionRequest struct {
	SongID uuid.UUID                      `validate:"required"`
	Name   string                         `validate:"required,max=30"`
	TypeID uuid.UUID                      `validate:"required"`
	Parts  []CreateSongSectionPartRequest `validate:"dive"`
}

type UpdateSongSectionRequest struct {
	ID      uuid.UUID `validate:"required"`
	Name    string    `validate:"required,max=30"`
	TypeID  uuid.UUID `validate:"required"`
	PartIDs []uuid.UUID
}

type MoveSongSectionRequest struct {
	ID     uuid.UUID `validate:"required"`
	OverID uuid.UUID `validate:"required"`
	SongID uuid.UUID `validate:"required"`
}

type BulkDeleteSongSectionsRequest struct {
	IDs     []uuid.UUID `validate:"unique,min=1"`
	SongID  uuid.UUID   `validate:"required"`
	PartIDs []uuid.UUID
}

type CreateSongSectionPartRequest struct {
	PartID       *uuid.UUID                `validate:"excluded_with=NewPart"`
	NewPart      *CreateNewSongPartRequest `validate:"excluded_with=PartID"`
	BandMemberID *uuid.UUID                `validate:"excluded_without_all=PartID NewPart"`
}

type CreateNewSongPartRequest struct {
	Name         string `validate:"required,max=30"`
	InstrumentID *uuid.UUID
}

File: repertoire.ui/src/features/songs/features/song/features/sections/types/requests/SongSectionRequests.ts
export interface GetSongSectionsRequest {
  songId: string
}

export interface CreateSongSectionRequest {
  songId: string
  typeId: string
  name: string
  parts: CreateNewSongSectionPartRequest[]
}

interface CreateNewSongSectionPartRequest {
  bandMemberId: string
  partId: string
  newPart: { name: string; instrumentId: string }
}

export interface UpdateSongSectionRequest {
  id: string
  typeId: string
  name: string
  partIds: string[]
}

export interface MoveSongSectionRequest {
  id: string
  overId: string
  songId: string
}

export interface BulkDeleteSongSectionsRequest {
  ids: string[]
  songId: string
}

export interface DeleteSongSectionRequest {
  id: string
  songId: string
  withParts?: boolean
}
